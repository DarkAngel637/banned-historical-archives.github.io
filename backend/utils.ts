export function normalize(path: string) {
  return path.replace(/\\/g, '/');
}