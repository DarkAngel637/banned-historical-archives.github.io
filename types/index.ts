import type Publication from '../backend/entity/publication';

export enum ArticleCategory {
  centralFile = '中央文件',
  editorial = '重要报刊和社论',
  keyFigures = '关键人物文稿',
  keyPapersFromTheMasses = '群众运动重要文献',
}

export enum TagType {
  articleCategory = '文稿大类',
  articleType = '文稿类型',
  place = '地点',
  character = '人物',
  issuer = '发行机构',
  subject = '主题',
}

export enum ContentType {
  appellation = 'appellation',
  title = 'title',
  subtitle = 'subtitle',
  subdate = 'subdate',
  paragraph = 'paragraph',
  quotation = 'quotation',
}

export type ContentPartRaw = {
  text: string;
  type: ContentType;
};
export type ContentPart = {
  text: string;
  type: ContentType;
};

export type Date = {
  year?: number;
  month?: number;
  day?: number;
};

export type Pivot = {
  part_idx: number; // 从 0 开始
  index: number; // 注释编号
  offset: number; // 偏移量，从 0 开始，注释应该插入的index，比如'mzd[2]'的offset为3
}

export type ParserResult = {
  title: string;
  alias?: string;
  dates: Date[];
  is_range_date: boolean;
  authors: string[];
  parts: ContentPart[];
  comments: string[];
  comment_pivots: Pivot[];
  description: string;
  page_start: number;
  page_end: number;
  origin?: string; // 起源
  tags?: {
    name: string,
    type: TagType
  }[];
  file_id?: string;

  title_raw?: string;
  date_raw?: string;
  parts_raw?: ContentPartRaw[];
};

export enum ArticleType {
  writings = '文章',
  mail = '书信',
  lecture = '发言',
  talk = '对话',
  declaration = '宣言',
  instruction = '指示',
  comment = '批示',
  telegram = '通讯',
}

export type OCRPosition = [number, number];
export type OCRResult = {
  // 坐标轴原点在左上角，y轴朝下
  // 左上，右上，右下，左下
  box: [OCRPosition, OCRPosition, OCRPosition, OCRPosition],
  text: string,
};
export type LACType =
  | 'n' // 普通名词
  | 'f' // 方位名词
  | 's' // 处所名词
  | 't' // 时间
  | 'nr' // 人名
  | 'ns' // 地名
  | 'nt' // 机构名
  | 'nw' // 作品名
  | 'nz' // 其他专名
  | 'v' // 普通动词
  | 'vd' // 动副词
  | 'vn' // 名动词
  | 'a' // 形容词
  | 'ad' // 副形词
  | 'an' // 名形词
  | 'd' // 副词
  | 'm' // 数量词
  | 'q' // 量词
  | 'r' // 代词
  | 'p' // 介词
  | 'c' // 连词
  | 'u' // 助词
  | 'xc' // 其他虚词
  | 'w' // 标点符号
  | 'PER' // 人名
  | 'LOC' // 地名
  | 'ORG' // 机构名
  | 'TIME'; // 时间

export type LACResult = {
  text: string;
  count: number;
  type: LACType;
};

export type ParserOption = {
  page_limits: [number, number][];
  page_width?: number;
  content_min_x?: number;
  name?: string;
  header_min_height?: number;
};

export type Book = {
  entity: Partial<Publication>;
  path: string;
  parser_option: ParserOption;
  parser: (path: string, opt: ParserOption) => Promise<ParserResult[]>;
};

export type Patch = {
  parts: {[idx: string]: string},
  comments: {[idx: string]: string},
  description: string,
}

export type StringDiff = string;

export type PartDiff = {
  insertBefore?: ContentPart[];
  insertAfter?: ContentPart[];
  delete?: boolean;
  diff?: StringDiff;
  type?: ContentType;
};

export type CommentDiff = {
  insertBefore?: { id?: string; text: StringDiff }[];
  insertAfter?: { id?: string; text: StringDiff }[];
  delete?: boolean;
  diff?: StringDiff;
};

export type PatchV2 = {
  version: 2,
  parts: {[idx: string]: PartDiff},
  comments: {[idx: string]: CommentDiff},
  newComments?: string[],
  description?: StringDiff, // 如果为空字符串表示无变更，如果不存在，表示删除
}