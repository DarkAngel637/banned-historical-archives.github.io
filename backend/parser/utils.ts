import { OCRResult } from "../../types";
import type { Item, ContentObj } from './pdf.js';

export function merge_to_lines(ocrResults: OCRResult[], threshold = 50) {
  const next = new Map<OCRResult, OCRResult>();
  const to_remove = new Map<OCRResult, boolean>();
  for (const a of ocrResults) {
    for (const b of ocrResults) {
      if (a == b) continue;
      if (next.get(a)) {
        continue;
      }
      const x_diff = Math.abs(b.box[0][0] - a.box[1][0]);
      const y_diff = Math.abs(b.box[0][1] - a.box[1][1]);
      if (x_diff + y_diff < threshold && !next.get(b)) {
        next.set(a, b);
        to_remove.set(b, true);
      }
    }
  }
  const lines: OCRResult[] = ocrResults
    .filter((i) => !to_remove.get(i))
    .map((i) => {
      let j = next.get(i);
      const r = i;
      while (j) {
        r.text += j.text;
        j = next.get(j);
      }
      return r;
    })
    .sort((a, b) => a.box[0][1] - b.box[0][1]);
  return lines;
}

export function pdfjsContentToOCRResult(obj: ContentObj, height: number):OCRResult[] {
  return obj.items
    .map((i) => {
      const x = i.transform[4];
      const y = height - i.transform[5];
      const r: OCRResult = {
        text: i.str,
        box: [
          [x, y],
          [x + i.width, y],
          [x + i.width, y + i.height],
          [x, y + i.height],
        ],
      };
      return r;
    })
    .filter((i) => i.text.length);
}