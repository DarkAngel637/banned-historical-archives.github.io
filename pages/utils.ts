import * as Diff from 'diff';
import { DiffResult, LineDiff } from "../types";

export function diff(
  contents_a: string[],
  contents_b: string[],
) {
  let i = 0;
  const a_diff: DiffResult = [];
  while (i < contents_a.length && i < contents_b.length) {
    const line_diffs: LineDiff[] = [];
    a_diff.push({ line_diffs, id: Math.random().toString() });
    const line_diff = Diff.diffChars(contents_a[i], contents_b[i]);
    for (const r of line_diff) {
      line_diffs.push({
        id: Math.random().toString(),
        removed: r.removed,
        added: r.added,
        value: r.value,
      });
    }
    ++i;
  }
  if (contents_a.length != contents_b.length) {
    const max_len = Math.max(contents_a.length, contents_b.length);
    const target =
      contents_a.length > contents_b.length ? contents_a : contents_b;
    while (i < max_len) {
      a_diff.push({
        line_diffs: [
          {
            id: Math.random().toString(),
            removed: target === contents_a ? false : true,
            added: target === contents_a ? true : false,
            value: target[i],
          },
        ],
        id: Math.random().toString(),
      });
      ++i;
    }
  }
  return a_diff;
}