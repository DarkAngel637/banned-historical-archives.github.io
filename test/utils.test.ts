import { assert, expect, test, it, describe, beforeEach, beforeAll } from 'vitest';
import { join } from 'node:path';
import { extract_dates } from '../backend/parser/utils';
import { ContentType, ParserResult } from '../types';
import { apply_patch_v2, bracket_left, bracket_right } from '../utils';

describe('extract_dates', async () => {
  it('format a', () => {
    expect(extract_dates('1911.10.10-1912.12.12')).toMatchSnapshot();
    expect(extract_dates('1911.10.10-12.12')).toMatchSnapshot();
    expect(extract_dates('1911.10.10-12')).toMatchSnapshot();
    expect(extract_dates('1911.10-12')).toMatchSnapshot();
  });
  it('format b', () => {
    expect(extract_dates('1911.10.10,11,12')).toMatchSnapshot();
    expect(extract_dates('1911.10,11,12')).toMatchSnapshot();
    expect(extract_dates('1911.10.21,10.22,11.23')).toMatchSnapshot();
    expect(extract_dates('1962.1.10,24,2.23')).toMatchSnapshot();
  });
  it('format c', () => {
    expect(extract_dates('1911.10.12')).toMatchSnapshot();
    expect(extract_dates('1911.10')).toMatchSnapshot();
    expect(extract_dates('1911')).toMatchSnapshot();
  });
  it('format d', () => {
    expect(extract_dates('一九三二年十月二十日至一九三三年十月二十日')).toMatchSnapshot();
    expect(extract_dates('一九三二年十月二十日至十二月二十日')).toMatchSnapshot();
    expect(extract_dates('一九三二年十月二十日至十二月')).toMatchSnapshot();
    expect(extract_dates('一九三二年十月至十二月')).toMatchSnapshot();
  });
  it('format e', () => {
    expect(extract_dates('一九三二年十月二十日，二十一日，二十二日')).toMatchSnapshot();
    expect(extract_dates('一九三二年十月二十日，十一月十日，十一月十二日')).toMatchSnapshot();
    expect(extract_dates('一九三二年十月，十一月，十二月')).toMatchSnapshot();
    expect(extract_dates('1962年1月10日、24日，2月23日')).toMatchSnapshot();
  });
  it('format f', () => {
    expect(extract_dates('一九三二年十月二十日')).toMatchSnapshot();
    expect(extract_dates('一九三二年十月')).toMatchSnapshot();
    expect(extract_dates('一九三二年')).toMatchSnapshot();
  });
  it('format a with remove_unknowns', () => {
    expect(
      extract_dates(':191:1.1:0.10-1912.12.1:2', { remove_unknowns: true }),
    ).toMatchSnapshot();
  });
});

describe('apply_patch_v2', async () => {
  const result: ParserResult = {
    title: 'test',
    dates: [],
    is_range_date: false,
    parts: [
      {
        type: ContentType.paragraph,
        text: `mzd111jq222ywy`,
      },
      {
        type: ContentType.paragraph,
        text: `ZCQ`,
      },
    ],
    page_start: 0,
    page_end: 0,
    authors: ['test'],
    comments: ['MZD', 'JQ', 'ZCQ'],
    comment_pivots: [
      {
        part_idx: 0,
        index: 1,
        offset: 3,
      },
      {
        part_idx: 0,
        index: 2,
        offset: 8,
      },
      {
        part_idx: 1,
        index: 3,
        offset: 3,
      },
    ],
    description: 'asdf',
  };
  it('part modify', () => {
    expect(
      apply_patch_v2(result, {
        version: 2,
        parts: {
          '0': {
            insertAfter: [],
            insertBefore: [],
            diff: '=20\t+1',
          },
        },
        comments: {},
        description: '',
      }),
    ).toMatchSnapshot();
  });
  it('pivot modify', () => {
    expect(
      apply_patch_v2(result, {
        version: 2,
        parts: {
          '0': {
            insertAfter: [],
            insertBefore: [],
            diff: '=4\t+5\t=16',
          },
        },
        comments: {},
        description: '',
      }),
    ).toMatchSnapshot();
  });
  it('pivot modify2', () => {
    expect(
      apply_patch_v2(result, {
        version: 2,
        parts: {
          '0': {
            insertAfter: [],
            insertBefore: [],
            diff: '=4\t+5\t=9\t+6\t=7',
          },
        },
        comments: {},
        description: '',
      }),
    ).toMatchSnapshot();
  });
  it('part insert before + insert after', () => {
    expect(
      apply_patch_v2(result, {
        version: 2,
        parts: {
          '0': {
            insertBefore: [{ type: ContentType.title, text: 'title' }],
            insertAfter: [{ type: ContentType.paragraph, text: 'signature' }],
          },
        },
        comments: {},
        description: '',
      }),
    ).toMatchSnapshot();
  });
  it('part insert before + insert after + delete', () => {
    expect(
      apply_patch_v2(result, {
        version: 2,
        parts: {
          '0': {
            insertBefore: [{ type: ContentType.title, text: 'title' }],
            insertAfter: [{ type: ContentType.paragraph, text: 'signature' }],
            delete: true,
          },
        },
        comments: {},
        description: '',
      }),
    ).toMatchSnapshot();
  });
  it('part insert before + insert after + pivot modification', () => {
    expect(
      apply_patch_v2(result, {
        version: 2,
        parts: {
          '0': {
            insertBefore: [{ type: ContentType.title, text: `title${bracket_left}23${bracket_right}` }],
            insertAfter: [{ type: ContentType.paragraph, text: 'signature' }],
          },
        },
        comments: {},
        description: '',
      }),
    ).toMatchSnapshot();
  });
  it('comment modify + insert before after', () => {
    expect(
      apply_patch_v2(result, {
        parts: {},
        version: 2,
        comments: {
          '1': {
            insertAfter: [{text: 'after1'}],
            insertBefore: [{text: 'before1'}],
            diff: '=2\t+XX\t=1',
          }
        },
        description: '',
      }),
    ).toMatchSnapshot();
  });
  it('comment delete', () => {
    expect(
      apply_patch_v2(result, {
        version: 2,
        parts: {},
        comments: {
          '2': {
            delete: true,
          }
        },
        description: '',
      }),
    ).toMatchSnapshot();
  });
  it('new comments', () => {
    expect(
      apply_patch_v2(result, {
        version: 2,
        parts: {},
        comments: {},
        newComments: ['a', 'b', 'c'],
        description: '',
      }),
    ).toMatchSnapshot();
  });
  it('description remove', () => {
    expect(
      apply_patch_v2(result, {
        version: 2,
        parts: {},
        comments: {},
      }),
    ).toMatchSnapshot();
  });
  it('description modification', () => {
    expect(
      apply_patch_v2(result, {
        version: 2,
        parts: {},
        comments: {},
        description: '=2\t+111\t=2'
      }),
    ).toMatchSnapshot();
  });
  it('type modification', () => {
    expect(
      apply_patch_v2(result, {
        version: 2,
        parts: {
          '0': {
            type: ContentType.subtitle3,
          },
        },
        comments: {},
        description: '',
      }),
    ).toMatchSnapshot();
  });
});