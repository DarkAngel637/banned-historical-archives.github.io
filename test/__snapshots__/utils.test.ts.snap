// Vitest Snapshot v1

exports[`apply_patch_v2 > comment delete 1`] = `
{
  "authors": [
    "test",
  ],
  "comment_pivots": [
    {
      "index": 1,
      "offset": 3,
      "part_idx": 0,
    },
    {
      "index": 2,
      "offset": 8,
      "part_idx": 0,
    },
    {
      "index": 3,
      "offset": 3,
      "part_idx": 1,
    },
  ],
  "comments": [
    "MZD",
    "ZCQ",
  ],
  "dates": [],
  "description": "asdf",
  "is_range_date": false,
  "page_end": 0,
  "page_start": 0,
  "parts": [
    {
      "text": "mzd111jq222ywy",
      "type": "paragraph",
    },
    {
      "text": "ZCQ",
      "type": "paragraph",
    },
  ],
  "title": "test",
}
`;

exports[`apply_patch_v2 > comment modify + insert before after 1`] = `
{
  "authors": [
    "test",
  ],
  "comment_pivots": [
    {
      "index": 1,
      "offset": 3,
      "part_idx": 0,
    },
    {
      "index": 2,
      "offset": 8,
      "part_idx": 0,
    },
    {
      "index": 3,
      "offset": 3,
      "part_idx": 1,
    },
  ],
  "comments": [
    "before1",
    "MZXXD",
    "after1",
    "JQ",
    "ZCQ",
  ],
  "dates": [],
  "description": "asdf",
  "is_range_date": false,
  "page_end": 0,
  "page_start": 0,
  "parts": [
    {
      "text": "mzd111jq222ywy",
      "type": "paragraph",
    },
    {
      "text": "ZCQ",
      "type": "paragraph",
    },
  ],
  "title": "test",
}
`;

exports[`apply_patch_v2 > description modification 1`] = `
{
  "authors": [
    "test",
  ],
  "comment_pivots": [
    {
      "index": 1,
      "offset": 3,
      "part_idx": 0,
    },
    {
      "index": 2,
      "offset": 8,
      "part_idx": 0,
    },
    {
      "index": 3,
      "offset": 3,
      "part_idx": 1,
    },
  ],
  "comments": [
    "MZD",
    "JQ",
    "ZCQ",
  ],
  "dates": [],
  "description": "as111df",
  "is_range_date": false,
  "page_end": 0,
  "page_start": 0,
  "parts": [
    {
      "text": "mzd111jq222ywy",
      "type": "paragraph",
    },
    {
      "text": "ZCQ",
      "type": "paragraph",
    },
  ],
  "title": "test",
}
`;

exports[`apply_patch_v2 > description remove 1`] = `
{
  "authors": [
    "test",
  ],
  "comment_pivots": [
    {
      "index": 1,
      "offset": 3,
      "part_idx": 0,
    },
    {
      "index": 2,
      "offset": 8,
      "part_idx": 0,
    },
    {
      "index": 3,
      "offset": 3,
      "part_idx": 1,
    },
  ],
  "comments": [
    "MZD",
    "JQ",
    "ZCQ",
  ],
  "dates": [],
  "description": "",
  "is_range_date": false,
  "page_end": 0,
  "page_start": 0,
  "parts": [
    {
      "text": "mzd111jq222ywy",
      "type": "paragraph",
    },
    {
      "text": "ZCQ",
      "type": "paragraph",
    },
  ],
  "title": "test",
}
`;

exports[`apply_patch_v2 > new comments 1`] = `
{
  "authors": [
    "test",
  ],
  "comment_pivots": [
    {
      "index": 1,
      "offset": 3,
      "part_idx": 0,
    },
    {
      "index": 2,
      "offset": 8,
      "part_idx": 0,
    },
    {
      "index": 3,
      "offset": 3,
      "part_idx": 1,
    },
  ],
  "comments": [
    "a",
    "b",
    "c",
  ],
  "dates": [],
  "description": "asdf",
  "is_range_date": false,
  "page_end": 0,
  "page_start": 0,
  "parts": [
    {
      "text": "mzd111jq222ywy",
      "type": "paragraph",
    },
    {
      "text": "ZCQ",
      "type": "paragraph",
    },
  ],
  "title": "test",
}
`;

exports[`apply_patch_v2 > part insert before + insert after + delete 1`] = `
{
  "authors": [
    "test",
  ],
  "comment_pivots": [
    {
      "index": 3,
      "offset": 3,
      "part_idx": 2,
    },
  ],
  "comments": [
    "MZD",
    "JQ",
    "ZCQ",
  ],
  "dates": [],
  "description": "asdf",
  "is_range_date": false,
  "page_end": 0,
  "page_start": 0,
  "parts": [
    {
      "text": "title",
      "type": "title",
    },
    {
      "text": "signature",
      "type": "paragraph",
    },
    {
      "text": "ZCQ",
      "type": "paragraph",
    },
  ],
  "title": "test",
}
`;

exports[`apply_patch_v2 > part insert before + insert after + pivot modification 1`] = `
{
  "authors": [
    "test",
  ],
  "comment_pivots": [
    {
      "index": 23,
      "offset": 5,
      "part_idx": 0,
    },
    {
      "index": 3,
      "offset": 3,
      "part_idx": 3,
    },
  ],
  "comments": [
    "MZD",
    "JQ",
    "ZCQ",
  ],
  "dates": [],
  "description": "asdf",
  "is_range_date": false,
  "page_end": 0,
  "page_start": 0,
  "parts": [
    {
      "text": "title",
      "type": "title",
    },
    {
      "text": "mzd111jq222ywy",
      "type": "paragraph",
    },
    {
      "text": "signature",
      "type": "paragraph",
    },
    {
      "text": "ZCQ",
      "type": "paragraph",
    },
  ],
  "title": "test",
}
`;

exports[`apply_patch_v2 > part insert before + insert after 1`] = `
{
  "authors": [
    "test",
  ],
  "comment_pivots": [
    {
      "index": 3,
      "offset": 3,
      "part_idx": 3,
    },
  ],
  "comments": [
    "MZD",
    "JQ",
    "ZCQ",
  ],
  "dates": [],
  "description": "asdf",
  "is_range_date": false,
  "page_end": 0,
  "page_start": 0,
  "parts": [
    {
      "text": "title",
      "type": "title",
    },
    {
      "text": "mzd111jq222ywy",
      "type": "paragraph",
    },
    {
      "text": "signature",
      "type": "paragraph",
    },
    {
      "text": "ZCQ",
      "type": "paragraph",
    },
  ],
  "title": "test",
}
`;

exports[`apply_patch_v2 > part modify 1`] = `
{
  "authors": [
    "test",
  ],
  "comment_pivots": [
    {
      "index": 1,
      "offset": 3,
      "part_idx": 0,
    },
    {
      "index": 2,
      "offset": 8,
      "part_idx": 0,
    },
    {
      "index": 3,
      "offset": 3,
      "part_idx": 1,
    },
  ],
  "comments": [
    "MZD",
    "JQ",
    "ZCQ",
  ],
  "dates": [],
  "description": "asdf",
  "is_range_date": false,
  "page_end": 0,
  "page_start": 0,
  "parts": [
    {
      "text": "mzd111jq222ywy1",
      "type": "paragraph",
    },
    {
      "text": "ZCQ",
      "type": "paragraph",
    },
  ],
  "title": "test",
}
`;

exports[`apply_patch_v2 > pivot modify 1`] = `
{
  "authors": [
    "test",
  ],
  "comment_pivots": [
    {
      "index": 51,
      "offset": 3,
      "part_idx": 0,
    },
    {
      "index": 2,
      "offset": 8,
      "part_idx": 0,
    },
    {
      "index": 3,
      "offset": 3,
      "part_idx": 1,
    },
  ],
  "comments": [
    "MZD",
    "JQ",
    "ZCQ",
  ],
  "dates": [],
  "description": "asdf",
  "is_range_date": false,
  "page_end": 0,
  "page_start": 0,
  "parts": [
    {
      "text": "mzd111jq222ywy",
      "type": "paragraph",
    },
    {
      "text": "ZCQ",
      "type": "paragraph",
    },
  ],
  "title": "test",
}
`;

exports[`apply_patch_v2 > pivot modify2 1`] = `
{
  "authors": [
    "test",
  ],
  "comment_pivots": [
    {
      "index": 51,
      "offset": 3,
      "part_idx": 0,
    },
    {
      "index": 26,
      "offset": 8,
      "part_idx": 0,
    },
    {
      "index": 3,
      "offset": 3,
      "part_idx": 1,
    },
  ],
  "comments": [
    "MZD",
    "JQ",
    "ZCQ",
  ],
  "dates": [],
  "description": "asdf",
  "is_range_date": false,
  "page_end": 0,
  "page_start": 0,
  "parts": [
    {
      "text": "mzd111jq222ywy",
      "type": "paragraph",
    },
    {
      "text": "ZCQ",
      "type": "paragraph",
    },
  ],
  "title": "test",
}
`;

exports[`extract_dates > format a 1`] = `
{
  "dates": [
    {
      "day": 10,
      "month": 10,
      "year": 1911,
    },
    {
      "day": 12,
      "month": 12,
      "year": 1912,
    },
  ],
  "is_range_date": true,
}
`;

exports[`extract_dates > format a 2`] = `
{
  "dates": [
    {
      "day": 10,
      "month": 10,
      "year": 1911,
    },
    {
      "day": 12,
      "month": 12,
      "year": 1911,
    },
  ],
  "is_range_date": true,
}
`;

exports[`extract_dates > format a 3`] = `
{
  "dates": [
    {
      "day": 10,
      "month": 10,
      "year": 1911,
    },
    {
      "day": 12,
      "month": 10,
      "year": 1911,
    },
  ],
  "is_range_date": true,
}
`;

exports[`extract_dates > format a 4`] = `
{
  "dates": [
    {
      "day": undefined,
      "month": 10,
      "year": 1911,
    },
    {
      "day": undefined,
      "month": 12,
      "year": 1911,
    },
  ],
  "is_range_date": true,
}
`;

exports[`extract_dates > format a with remove_unknowns 1`] = `
{
  "dates": [
    {
      "day": 10,
      "month": 10,
      "year": 1911,
    },
    {
      "day": 12,
      "month": 12,
      "year": 1912,
    },
  ],
  "is_range_date": true,
}
`;

exports[`extract_dates > format b 1`] = `
{
  "dates": [
    {
      "day": 10,
      "month": 10,
      "year": 1911,
    },
    {
      "day": 11,
      "month": 10,
      "year": 1911,
    },
    {
      "day": 12,
      "month": 10,
      "year": 1911,
    },
  ],
  "is_range_date": false,
}
`;

exports[`extract_dates > format b 2`] = `
{
  "dates": [
    {
      "day": undefined,
      "month": 10,
      "year": 1911,
    },
    {
      "day": undefined,
      "month": 11,
      "year": 1911,
    },
    {
      "day": undefined,
      "month": 12,
      "year": 1911,
    },
  ],
  "is_range_date": false,
}
`;

exports[`extract_dates > format b 3`] = `
{
  "dates": [
    {
      "day": 21,
      "month": 10,
      "year": 1911,
    },
    {
      "day": 22,
      "month": 10,
      "year": 1911,
    },
    {
      "day": 23,
      "month": 11,
      "year": 1911,
    },
  ],
  "is_range_date": false,
}
`;

exports[`extract_dates > format b 4`] = `
{
  "dates": [
    {
      "day": 10,
      "month": 1,
      "year": 1962,
    },
    {
      "day": 24,
      "month": 1,
      "year": 1962,
    },
    {
      "day": 23,
      "month": 2,
      "year": 1962,
    },
  ],
  "is_range_date": false,
}
`;

exports[`extract_dates > format c 1`] = `
{
  "dates": [
    {
      "day": 12,
      "month": 10,
      "year": 1911,
    },
  ],
  "is_range_date": false,
}
`;

exports[`extract_dates > format c 2`] = `
{
  "dates": [
    {
      "day": undefined,
      "month": 10,
      "year": 1911,
    },
  ],
  "is_range_date": false,
}
`;

exports[`extract_dates > format c 3`] = `
{
  "dates": [
    {
      "day": undefined,
      "month": undefined,
      "year": undefined,
    },
  ],
  "is_range_date": false,
}
`;

exports[`extract_dates > format d 1`] = `
{
  "dates": [
    {
      "day": 20,
      "month": 10,
      "year": 1932,
    },
    {
      "day": 20,
      "month": 10,
      "year": 1933,
    },
  ],
  "is_range_date": true,
}
`;

exports[`extract_dates > format d 2`] = `
{
  "dates": [
    {
      "day": 20,
      "month": 10,
      "year": 1932,
    },
    {
      "day": 20,
      "month": 12,
      "year": 1932,
    },
  ],
  "is_range_date": true,
}
`;

exports[`extract_dates > format d 3`] = `
{
  "dates": [
    {
      "day": 20,
      "month": 10,
      "year": 1932,
    },
    {
      "day": 12,
      "month": 10,
      "year": 1932,
    },
  ],
  "is_range_date": true,
}
`;

exports[`extract_dates > format d 4`] = `
{
  "dates": [
    {
      "day": undefined,
      "month": 10,
      "year": 1932,
    },
    {
      "day": undefined,
      "month": 12,
      "year": 1932,
    },
  ],
  "is_range_date": true,
}
`;

exports[`extract_dates > format e 1`] = `
{
  "dates": [
    {
      "day": 20,
      "month": 10,
      "year": 1932,
    },
    {
      "day": 21,
      "month": 10,
      "year": 1932,
    },
    {
      "day": 22,
      "month": 10,
      "year": 1932,
    },
  ],
  "is_range_date": false,
}
`;

exports[`extract_dates > format e 2`] = `
{
  "dates": [
    {
      "day": 20,
      "month": 10,
      "year": 1932,
    },
    {
      "day": 10,
      "month": 11,
      "year": 1932,
    },
    {
      "day": 12,
      "month": 11,
      "year": 1932,
    },
  ],
  "is_range_date": false,
}
`;

exports[`extract_dates > format e 3`] = `
{
  "dates": [
    {
      "day": undefined,
      "month": 10,
      "year": 1932,
    },
    {
      "day": undefined,
      "month": 11,
      "year": 1932,
    },
    {
      "day": undefined,
      "month": 12,
      "year": 1932,
    },
  ],
  "is_range_date": false,
}
`;

exports[`extract_dates > format e 4`] = `
{
  "dates": [
    {
      "day": 10,
      "month": 1,
      "year": 1962,
    },
    {
      "day": 24,
      "month": 1,
      "year": 1962,
    },
    {
      "day": 23,
      "month": 2,
      "year": 1962,
    },
  ],
  "is_range_date": false,
}
`;

exports[`extract_dates > format f 1`] = `
{
  "dates": [
    {
      "day": 20,
      "month": 10,
      "year": 1932,
    },
  ],
  "is_range_date": false,
}
`;

exports[`extract_dates > format f 2`] = `
{
  "dates": [
    {
      "day": undefined,
      "month": 10,
      "year": 1932,
    },
  ],
  "is_range_date": false,
}
`;

exports[`extract_dates > format f 3`] = `
{
  "dates": [
    {
      "day": undefined,
      "month": undefined,
      "year": 1932,
    },
  ],
  "is_range_date": false,
}
`;
